frontend/src/components/PhoneNumberInput/PhoneNumberInput.tsx

```diff
@@ -0,0 +1,304 @@
+   * blurred.
+   */
+  onBlur?: () => void
+  /**
+   * The current value of the controlled component.
+   */
+  value: string | undefined
+
+  /**
+   * Set the input's placeholder to an example number for the selected country,
+   * and update it if the country changes.
+   *
+   * By default it is set to "polite", which means it will only set the
+   * placeholder if the input doesn't already have one. You can also set it to
+   * "aggressive", which will replace any existing placeholder, or "off" to not
+   * show any example numbers in the placeholder.
+   */
+  examplePlaceholder?: 'polite' | 'aggressive' | 'off'
+
+  /**
+   * Examples to retrieve placeholder number from, if any. Defaults to
+   * `libphonenumber-js/examples.mobile.json` if none provided.
+   */
+  examples?: { [country in CountryCode]: NationalNumber }
+}
+
+export const PhoneNumberInput = forwardRef<PhoneNumberInputProps, 'input'>(
+  (
+    {
+      defaultCountry = 'SG',
+      onChange,
+      onBlur,
+      value: propsValue,
+      isDisabled,
+      examples = defaultExamples,
+      examplePlaceholder = 'polite',
+      ...props
+    },
+    ref,
+  ) => {
+    const styles = useMultiStyleConfig('PhoneNumberInput', props)
+
+    // Internal states of the component.
+    const [inputValue, setInputValue] = useState(propsValue ?? '')
+    const [country, setCountry] = useState(defaultCountry)
+
+    // Refs of the phone number input so focus can be passed to the input when
+    // the selected country changes.
+    const innerInputRef = useRef<HTMLInputElement | null>(null)
+    // Used so any forwarded refs passed can be merged with internal ref.
+    const inputRef = useMergeRefs(innerInputRef, ref)
+
+    const formatter = useMemo(() => new AsYouType(country), [country])
+
+    const inputPlaceholder = useMemo(() => {
+      if (examplePlaceholder === 'off') {
+        return props.placeholder
+      }
+
+      const exampleNumber = getExampleNumber(
+        country,
+        examples,
+      )?.formatNational()
+
+      if (examplePlaceholder === 'aggressive') {
+        return exampleNumber ?? props.placeholder
+      }
+
+      return props.placeholder ?? exampleNumber
+    }, [country, examplePlaceholder, examples, props.placeholder])
+
+    const onInputChange = useCallback(
+      (newValue: string) => {
+        if (inputValue === newValue) return
+
+        // The as-you-type formatter only works with append-only inputs.
+        // Changes other than append require a reset.
+        const isAppend =
+          newValue.length > inputValue.length &&
+          newValue.slice(0, inputValue.length) === inputValue
+
+        if (isAppend) {
+          const appended = newValue.slice(inputValue.length)
+          setInputValue(formatter.input(appended))
+
+          const number = formatter.getNumber()
+          if (number?.country && number.country !== country) {
+            setCountry(number.country)
+          }
```

Comment:
could we add a comment here fpr what this does?

Reply:
this is just setting the country state with the processed number~